src/compute-client/src/controller/replicated.rs

```diff
@@ -248,81 +348,115 @@ where
+                                            lower: new_lower,
+                                            upper: entry.bounds.upper.clone(),
+                                            updates,
+                                        }),
+                                    ),
+                                )
+                            }
+
+                            let frontier_updates = vec![(id, entry.bounds.clone())];
+                            self.pending_response.push_back(
+                                ActiveReplicationResponse::FrontierUppers(frontier_updates),
+                            );
                         }
                     }
                     TailResponse::DroppedAt(frontier) => {
-                        // Introduce a new terminal frontier to suppress all future responses.
-                        // We cannot simply remove the entry, as we currently create new entries in response
-                        // to observed responses; if we pre-load the entries in response to commands we can
-                        // clean up the state here.
-                        self.tails.insert(id, Antichain::new());
-                        Some(ActiveReplicationResponse::ComputeResponse(
-                            ComputeResponse::TailResponse(id, TailResponse::DroppedAt(frontier)),
-                        ))
+                        // Advance the tail upper to the empty frontier, to suppress all future
+                        // `TailResponse`s. We still need to keep tracking this tail, to ensure its
+                        // inputs don't get compacted as long as slower replicas might still be
+                        // reading from them. Only once all replicas have advanced to the empty
+                        // frontier, we can remove the entry.
+
+                        // Pass on only the first `DroppedAt` response we see for a tail.
+                        if !entry.bounds.upper.is_empty() {
+                            self.pending_response.push_back(
+                                ActiveReplicationResponse::TailResponse(
+                                    id,
+                                    TailResponse::DroppedAt(frontier),
+                                ),
+                            );
+                        }
+
+                        if entry.update(replica_id, Antichain::new()) {
+                            let frontier_updates = vec![(id, entry.bounds.clone())];
+                            self.pending_response.push_back(
+                                ActiveReplicationResponse::FrontierUppers(frontier_updates),
+                            );
+                        }
+
+                        // If all replicas have advanced to the empty frontier, stop tracking the
+                        // tail.
+                        if entry.bounds.lower.is_empty() {
+                            self.uppers.remove(&id);
+                        }
```

Comment:
If all but one replica has advanced to the empty frontier, and then we drop the remaining replica, I think we'll never clean up?

Reply:
You are right! I realized that tails are now covered by `frontier_tracking`, so we don't need to worry about cleaning up here at all. Instead cleanup works the same as for indexes and materialized views:

1. `uppers.bounds.lower` advances to the empty frontier (through some combination of frontier updates and replicas being dropped).
2. `ActiveReplication` produces a `FrontierUppers` response communicating that fact to the compute controller.
3. The compute controller produces an `AllowCompaction` command with the empty frontier for the tail collection.
4. The `AllowCompaction` command gets picked up by `frontier_tracking`, which tells `ActiveReplication` that we should cease tracking the tail frontier.
5. `ActiveReplication` cleans up the tail's entry.